pyresample/geometry.py

```diff
@@ -717,37 +717,282 @@ def _do_transform(src, dst, lons, lats, alt):
-                            self.lons.data, self.lats.data,
-                            da.zeros_like(self.lons.data), new_axis=[2],
-                            chunks=(self.lons.chunks[0], self.lons.chunks[1], 3))
-        res = DataArray(res, dims=['y', 'x', 'coord'], coords=self.lons.coords)
-        res = res.coarsen(**dims).mean()
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'], coords=src_lons.coords)
+
+        # Aggregating
+        res = res.coarsen(x=x, y=y, **kwargs).mean()
+
+        # Back-conversion to geographic CRS
         lonlatalt = da.map_blocks(self._do_transform, geocent, latlong,
-                                  res[:, :, 0].data, res[:, :, 1].data,
-                                  res[:, :, 2].data, new_axis=[2],
+                                  res[:, :, 0].data,  # x
+                                  res[:, :, 1].data,  # y
+                                  res[:, :, 2].data,  # z
+                                  new_axis=[2],
                                   chunks=res.data.chunks)
-        lons = DataArray(lonlatalt[:, :, 0], dims=self.lons.dims,
-                         coords=res.coords, attrs=self.lons.attrs.copy())
-        lats = DataArray(lonlatalt[:, :, 1], dims=self.lons.dims,
-                         coords=res.coords, attrs=self.lons.attrs.copy())
-        try:
-            resolution = lons.attrs['resolution'] * ((dims.get('x', 1) + dims.get('y', 1)) / 2)
-            lons.attrs['resolution'] = resolution
-            lats.attrs['resolution'] = resolution
-        except KeyError:
-            pass
+
+        # Back-conversion array as input format
+        lons, _ = _convert_2D_array(lonlatalt[:, :, 0], to=src_lons_format, dims=['y', 'x'])
+        lats, _ = _convert_2D_array(lonlatalt[:, :, 1], to=src_lats_format, dims=['y', 'x'])
+
+        # Add additional info if the source array is a DataArray
+        if isinstance(self.lats, xr.DataArray) and isinstance(self.lons, xr.DataArray):
+            lats = lats.assign_coords(res.coords)
+            lats.attrs = self.lats.attrs.copy()
+            lons = lons.assign_coords(res.coords)
+            lons.attrs = self.lons.attrs.copy()
+            try:
+                resolution = lons.attrs['resolution'] * ((x + y) / 2)
+                lons.attrs['resolution'] = resolution
+                lats.attrs['resolution'] = resolution
+            except KeyError:
+                pass
+
+        # Return the downsampled swath definition
+        return SwathDefinition(lons, lats)
+
+    def upsample(self, x=1, y=1):
+        """Upsample the swath definition along x (along-track) and y (cross-track) dimensions.
```

Comment:
Please don't use `along-track` and `accross-track`. Pyresample is use for many data type that are not satellite data. Prefer `lines` and `columns`.

Reply:
Done